RFC: Community plugins (#5610)
flyteorg/flyte: rfc/system/0008-community-plugins.md

# Management of community-contributed plugins

**Authors:**

- @davidmirror-ops

## 1 Executive Summary

The Flyte community as a self-governed and productive collective of individuals, values contributions. This proposal aims to discuss the process community contributors should follow to submit a new `flytekit` plugin, with special attention to mechanisms that ensure stability and maintainability of core flyte code.

## 2 Motivation

- With the current "in-tree" approach, plugins developed by the community land on the `flytekit` repo ([example](https://github.com/flyteorg/flytekit/pull/2537)). It results in Flyte maintainers having to take care of CI test failures due to plugin code or flytekit updates incompatible with plugin code, etc.

- The goal is to agree on a process for contributors to follow when submitting new integrations in a "out-of-tree" way that clearly communicates that it is a community-contributed -and then- community-supported integration.

## 3 Proposed Implementation

> - Create a `community` folder under `flytekit/plugins` and publish the plugins in that folder as separate `pypi` packages.

Reviewer: All plugins are separate pypi packages.

- Configure CI to only run tests on `plugins/community` when there are changes.
- Keep releasing plugins alongside flytekit.

> - Update [CODEOWNERS](https://github.com/flyteorg/flytekit/blob/master/CODEOWNERS), enabling contributors to merge community plugins (it requires granting Write permissions to contributors on `plugins/community`).

Reviewer: Is this possible in github?

Author: maybe I should rephrase it. It's possible to make some CODEOWNER on a folder but the Write permission is at the repo level AFAIK

> 

Reviewer: What do you think of explicitly marking them as community maintained in the import via `import flytekitplugins.contrib.x`?

Reviewer 2: How about `import flytekitplugins.community.x`?
like https://pypi.org/project/langchain-community/

Reviewer 3: I like it!

Reviewer 2: I believe we will have `flytekitplugins.experimental` and `flytekitplugins.partners` in the future.

Reviewer: `contrib` has been used e.g. by tensorflow back in the day.

Reviewer 2: yup. I'm okay with both

Author: I agree with using `contrib` as it communicates even more broadly, beyond the location in the repo

Author: cc @eapolinario

Reviewer 4: +1 to `contrib`.

### Promotion process to "core"/"official" plugin

@pryce-tuner [proposed criteria](https://github.com/flyteorg/flytekit/pull/2615/files#diff-c26d70784719ec01ce402e532a281a3e509d47523c3a1eba9b2c8d02933886a4R21) to promote plugins to be part of core flytekit, including the effects.

> 

Reviewer: How will the decision be taken? Vote by the steering committee?

Author: Yep, same as with any other RFC

Author: Sorry, I think you meant voting to promote a plugin, not the process itself. And I think it's also up to the TSC :) to arrive to a final decision to promote/not promote a plugin. This has to be added to the proposal

Reviewer 4: Plugin maintainers can also make the case that a plugin should be promoted, which will be evaluated by the TSC based on past stability, usage, etc.

## 3 Proposed Implementation

- Create a `community` folder under `flytekit/plugins` and keep releasing the plugins in that folder as separate `pypi` packages.
- Configure CI to only run tests on `plugins/community` when there are changes.
- Keep releasing plugins alongside flytekit.
- Update [CODEOWNERS](https://github.com/flyteorg/flytekit/blob/master/CODEOWNERS), enabling contributors to merge community plugins (it requires granting Write permissions to contributors on `flytekit`).

### Promotion process to "core"/"official" plugin

@pryce-tuner [proposed criteria](https://github.com/flyteorg/flytekit/pull/2615/files#diff-c26d70784719ec01ce402e532a281a3e509d47523c3a1eba9b2c8d02933886a4R21) to promote plugins to be part of core flytekit, including the effects.
- The supermajority of the TSC should approve publicly before promoting a plugin.

## 4 Drawbacks

- There's a management overhead associated with this initiative:
    - For every new plugin contributor, we may need to update `CODEOWNERS`

>     - CI configuration changes in flytekit (probably a one-time change) 

Reviewer 4: we'll need to rearrange a few things, but yeah, definitely a one-time small change.

## 5 Alternatives

- Maintain community plugins on a separate repo
    - Against the monorepo initiative
-  Have community packages be it's own org
    - Significantly higher management overhead
- `flytekit` plugins built into their own package
    -   Potentially heavier development process

## 6 Open questions

> - Does this apply to Agents?

Reviewer 2: it should. many people contribute agents as well

# Management of community-contributed plugins

**Authors:**

- @davidmirror-ops

## 1 Executive Summary

The Flyte community as a self-governed and productive collective of individuals, values contributions. This proposal aims to discuss the process community contributors should follow to submit a new `flytekit` plugin, with special attention to mechanisms that ensure stability and maintainability of core flyte code.

## 2 Motivation

> - With the current "in-tree" approach, plugins developed by the community land on the `flytekit` repo ([example](https://github.com/flyteorg/flytekit/pull/2537)). It results in Flyte maintainers having to take care of CI test failures due to plugin code or flytekit updates incompatible with plugin code, etc.

Reviewer: ```suggestion
- With the current "in-tree" approach, plugins developed by the community land in the `flytekit` repo ([example](https://github.com/flyteorg/flytekit/pull/2537)). It results in Flyte maintainers having to take care of CI test failures due to plugin code or flytekit updates incompatible with plugin code, etc.
```

- The goal is to agree on a process for contributors to follow when submitting new integrations in a "out-of-tree" way that clearly communicates that it is a community-contributed -and then- community-supported integration.

## 3 Proposed Implementation

- Create a `community` folder under `flytekit/plugins` and keep releasing the plugins in that folder as separate `pypi` packages.

> - Configure CI to only run tests on `plugins/community` when there are changes.

Reviewer: ```suggestion
- Configure CI to only run tests on `plugins/community` when there are changes to a respective plugin.
```

- Keep releasing plugins alongside flytekit.

> - Update [CODEOWNERS](https://github.com/flyteorg/flytekit/blob/master/CODEOWNERS) for the `plugins/community` folder, enabling contributors to merge community plugins (it requires granting Write permissions to contributors on `flytekit`).

Reviewer: Let's talk about this point today. Will every plugin author automatically get the permission to approve and merge PRs in the entire repo? Github's permissions model is a bit limiting for us here.

> - Update [CODEOWNERS](https://github.com/flyteorg/flytekit/blob/master/CODEOWNERS) for the `plugins/community` folder, enabling contributors to merge community plugins (it requires granting Write permissions to contributors on `flytekit`).

Author: Revisiting this: how the contribution process would look like?  Shouldn't be the existing CODEOWNERS (a.k.a. flytekit maintainers) who review and merge a new community plugin? In such case, we wouldn't need to update CODEOWNERS for every new plugin contribution.
Otherwise with the Github permission model, any new plugin contributor would have to be granted write permissions to the entire `flytekit` repo. 
cc @eapolinario

> - Update [CODEOWNERS](https://github.com/flyteorg/flytekit/blob/master/CODEOWNERS) for the `plugins/community` folder, enabling contributors to merge community plugins (it requires granting Write permissions to contributors on `flytekit`).

Reviewer: > Shouldn't be the existing CODEOWNERS (a.k.a. flytekit maintainers) who review and merge a new community plugin?

Yes, a new plugin should be merged by existing maintainers. The question for me is rather what happens after a plugin has been merged and needs to be maintained by the authors.

> - Update [CODEOWNERS](https://github.com/flyteorg/flytekit/blob/master/CODEOWNERS) for the `plugins/community` folder, enabling contributors to merge community plugins (it requires granting Write permissions to contributors on `flytekit`).

Author: What we talked with Eduardo today is that, considering the implication of CODEOWNERS, we would go the regular contribution path: plugin authors would have non-binding approval rights on changes to their plugins, and flytekit maintainers issue the final approvals and merge. We'd also keep releasing community plugins with flytekit, even if there are no changes. Does that make sense?

> - Plugin authors are responsible for maintaining their plugins, and `flytekit` maintainers are responsible for final approvals and merging into the repo.

Reviewer: ```suggestion
- Plugin authors are responsible for maintaining their plugins. In case there are PRs to change a community plugin, the plugin maintainers review the PR and give a non-binding approval. Once a community plugin maintainer has given a non-binding approval, a `flytekit` maintainer has to give a binding approval in order for the PR to be merged.
```

> 

Reviewer: ```suggestion

This proposal includes agent plugins.
```

Let's add this here so that we can remove this open question.

### Promotion process to official plugin

An official plugin is one that is maintained by the core Flyte team and is made part of the official `flytekit` documentation.

> - Plugins maintainers or community members can propose the promotion of a plugin to official by creating an Issue on the `flytekit` repo.

Reviewer: ```suggestion
- Plugin maintainers or community members can propose the promotion of a plugin to official by creating an Issue on the `flytekit` repo.
```

- The supermajority of the TSC must approve publicly before promoting a plugin.

To consider it for promotion, a plugin must meet the following criteria:

- Production readiness testing performed by the core Flyte team or documented by plugin users or maintainers
- Evidence of ongoing usage through Github issues or Slack threads

> 

Reviewer: ```suggestion
- Documented in flytekit's documentation
```

## 4 Drawbacks

- There's a management overhead associated with this initiative:
    - For every new plugin contributor, we may need to update `CODEOWNERS`
    - CI configuration changes in flytekit (probably a one-time change) 

## 5 Alternatives

- Maintain community plugins on a separate repo
    - Against the monorepo initiative
-  Have community packages be it's own org
    - Significantly higher management overhead
- `flytekit` plugins built into their own package
    -   Potentially heavier development process

## 6 Open questions

> 
> - Does this apply to Agents?

Reviewer: ```suggestion
-
```
